eclib/JWord_array.ec

```diff
@@ -250,8 +250,187 @@ abstract theory WArray.
+    proof gt0_sizeW by apply gt0_sizeWB, gt0_sizeA by apply gt0_sizeB.
+
+  op get_sub_direct (a: WordB.t ArrayWordsB.ArrayN.t) (i: int) =
+    ArrayWordsS.to_word_array (
+      ArrayWordsS.WArrayN.init (fun j =>
+        ArrayWordsB.WArrayN.get8 (ArrayWordsB.of_word_array a) (i + j)
+      )
+    ).
+
+  op set_sub_direct (a: WordB.t ArrayWordsB.ArrayN.t) (i: int) (b: WordS.t ArrayWordsS.ArrayN.t) =
+    (* FIXME: change init8 and get8 to init and _.[_] (breaks compatibility by rewrite) *)
+    ArrayWordsB.to_word_array (
+      ArrayWordsB.WArrayN.init8 (fun j =>
+        if i <= j < i + sizeWS * sizeS then
+          ArrayWordsS.WArrayN.get8 (ArrayWordsS.of_word_array b) (j - i)
+        else
+          ArrayWordsB.WArrayN.get8 (ArrayWordsB.of_word_array a) j
+        )
+      ).
+
+  op get_sub (a: WordB.t ArrayWordsB.ArrayN.t) (i: int) = get_sub_direct a (sizeWB*i).
+
+  op set_sub (a: WordB.t ArrayWordsB.ArrayN.t) (i: int) (b: WordS.t ArrayWordsS.ArrayN.t) =
+    set_sub_direct a (sizeWB*i) b.
+end SubArrayCast.
+
+(* Particular case of SubArrayCast where both array's word types are the same.
+Still not a simple sub-array since indices may be a number of bytes
+non-multiple of the word size. *)
+abstract theory SubArrayDirect.
+  op sizeW: int.
+  axiom gt0_sizeW: 0 < sizeW.
+
+  clone WT as Word with
+    op size <- 8 * sizeW
+    proof gt0_size by rewrite pmulr_rgt0 1:// gt0_sizeWS.
+
+  clone include SubArrayCast with
+    op sizeWS <- sizeW, op sizeWB <- sizeW, theory WordS <- Word, theory WordB <- Word
+    proof gt0_sizeWS by apply gt0_sizeW, gt0_sizeWB by apply gt0_sizeW.
+end SubArrayDirect.
+
+(* Access to array words (get, set) where the word may have a different type
+than the base array. Conversion views arrays of words and words as equivalent
+to byte arrays. *)
+abstract theory ArrayAccessCast.
+  op sizeWS: int.
+  axiom gt0_sizeWS: 0 < sizeWS.
+
+  op sizeWB: int.
+  axiom gt0_sizeWB: 0 < sizeWB.
+
+  op sizeB: int.
+  axiom gt0_sizeB: 0 < sizeB.
+
+  clone WT as WordS with
+    op size <- 8 * sizeWS
+    proof gt0_size by rewrite pmulr_rgt0 1:// gt0_sizeWS.
+
+  clone WT as WordB with
+    op size <- 8 * sizeWB
+    proof gt0_size by rewrite pmulr_rgt0 1:// gt0_sizeWB.
+
+  clone ArrayWords as ArrayWordsB with
+    op sizeW <- sizeWB, op sizeA <- sizeB, theory Word <- WordB
+    proof gt0_sizeW by apply gt0_sizeWB, gt0_sizeA by apply gt0_sizeB.
+
+  clone W_WS as WSu8 with
+    op sizeS <= W8.size, op sizeB <= W8.size*sizeWS, op r <= sizeWS,
+    theory WS <= W8, theory WB <= WordS
+    proof gt0_r by apply gt0_sizeWS, sizeBrS by rewrite mulzC.
+
+
+  op get_cast_direct (a: WordB.t ArrayWordsB.ArrayN.t) (i: int) =
+    WSu8.pack'R_t (
+      WSu8.Pack.init (fun j => ArrayWordsB.WArrayN.get8 (ArrayWordsB.of_word_array a) (i+j))
+    ).
+
+  op set_cast_direct (a: WordB.t ArrayWordsB.ArrayN.t) (i: int) (b: WordS.t) =
+    (* FIXME: change init8 and get8 to init and _.[_] (breaks compatibility by rewrite) *)
```

Comment:
Shall this be fixed before the PR is merged? Otherwise please change to TODO.

Reply:
The goal of not fixing this immediately was to allow to fix all proofs impacted by this change by simply `rewrite`'ing all new definitions. Here, the `rewrite` would need to go in the other direction, since get8 and init8 are just defined as _.[_] and init - or the proof could be fixed in a cleaner way (shouldn't be much harder in general).

Given that we have a backward-compatibility option, I'd tend to just make this additional change now.